src/pvnet_app/save.py

```diff
@@ -172,3 +180,151 @@ def convert_dataarray_to_forecasts(
+    list_forecasters_response = await client.list_forecasters(list_forecasters_request)
+    forecasters = list_forecasters_response.forecasters
+
+    forecasters_filtered = [f for f in forecasters if f.forecaster_name == name]
+
+    if len(forecasters_filtered) > 0:
+        forecaster = forecasters_filtered[0]
+    else:
+        cf_request = dp.CreateForecasterRequest(
+            name=name,
+            version=app_version,
+        )
+        forecaster_response = await client.create_forecaster(cf_request)
+        forecaster = forecaster_response.forecaster
+
+    # now loop over all gsps
+    for gsp_id in forecast_da.gsp_id.values:
+        logger.debug(f"Saving forecast for GSP ID: {gsp_id}")
+
+        # 3. get Location
+        location = uk_national_and_gsp_locations[int(gsp_id)]
+
+        # 4. Format the forecast values
+        forecast_values = get_forecast_values_from_dataarray(
+            forecast_da,
+            gsp_id=gsp_id,
+            init_time_utc=init_time_utc,
+            capacity_watts=location.effective_capacity_watts,
+        )
+        # 5. Save to data platform
+        forecast_request = dp.CreateForecastRequest(
+            forecaster=forecaster,
+            location_uuid=location.location_uuid,
+            energy_source=dp.EnergySource.SOLAR,
+            init_time_utc=init_time_utc.replace(tzinfo=UTC),
+            values=forecast_values,
+        )
+        _ = await client.create_forecast(forecast_request)
+
+
+def get_forecast_values_from_dataarray(
+    forecast_da: xr.DataArray,
+    gsp_id: int,
+    init_time_utc: datetime,
+    capacity_watts: int,
+) -> list[dp.CreateForecastRequestForecastValue]:
+    """Convert a DataArray for a single GSP to a list of ForecastValue objects.
+
+    Args:
+        forecast_da: DataArray for a single GSP
+        gsp_id: GSP ID
+        init_time_utc: Forecast initialization time
+        capacity_watts: Capacity of the location in watts
+
+    """
+    da_gsp = forecast_da.sel(gsp_id=gsp_id)
+
+    forecast_values = []
+    for target_time in pd.to_datetime(da_gsp.target_datetime_utc.values):
+        da_gsp_time = da_gsp.sel(target_datetime_utc=target_time)
+
+        target_time = target_time.replace(tzinfo=UTC)
+        horizon_mins = int((target_time - init_time_utc).total_seconds() / 60)
+        p50_fraction = da_gsp_time.sel(output_label="forecast_mw").item() * 10**6 / capacity_watts
+
+        # TODO tidy this up
+        metadata = Struct(fields={"temp": Value(string_value="temp")})
+
+        forecast_value = dp.CreateForecastRequestForecastValue(
+            horizon_mins=horizon_mins,
+            p50_fraction=p50_fraction,
+            metadata=metadata,
+            other_statistics_fractions={
+                "p10": 0.01,
+                "p90": 0.99,
+            },
+        )
+
+        forecast_values.append(forecast_value)
+
+    return forecast_values
+
+
+async def get_all_gsp_and_national_locations(
```

Comment:
Likewise everything in this function is awaited on so it also may not need to be async